packages/core/contracts/test/Storage.sol

```diff
@@ -431,4 +431,256 @@ contract StorageUpgrade_StructEnum_V2 {
         MyEnum myEnum;
     }
     MyStruct[] public structs;
-}
\ No newline at end of file
+}
+
+contract StorageUpgrade_Gap_V1 {
+    uint256 a;
+    uint256 b;
+    uint256[48] __gap;
+    uint256 z;
+}
+
+contract StorageUpgrade_Gap_V2_Ok {
+    uint256 a;
+    uint256 b;
+    uint256 c;
+    uint256[47] __gap;
+    uint256 z;
+}
+
+// insert var without shrink gap
+contract StorageUpgrade_Gap_V2_Bad1 {
+    uint256 a;
+    uint256 b;
+    uint256 c;
+    uint256[48] __gap;
+    uint256 z;
+}
+
+// delete var and expand gap
+contract StorageUpgrade_Gap_V2_Bad2 {
+    uint256 a;
+    uint256[49] __gap;
+    uint256 z;
+}
+
+// shrink gap without adding var
+contract StorageUpgrade_Gap_V2_Bad3 {
+    uint256 a;
+    uint256 b;
+    uint256[47] __gap;
+    uint256 z;
+}
+
+// insert var and shrink gap too much
+contract StorageUpgrade_Gap_V2_Bad4 {
+    uint256 a;
+    uint256 b;
+    uint256 c;
+    uint256[46] __gap;
+    uint256 z;
+}
+
+// insert vars and shrink gap not enough
+contract StorageUpgrade_Gap_V2_Bad5 {
+    uint256 a;
+    uint256 b;
+    uint256 c;
+    uint256 d;
+    uint256[47] __gap;
+    uint256 z;
+}
+
+contract StorageUpgrade_MultiConsumeGap_V1 {
+    uint256[2] __gap;
+}
+
+contract StorageUpgrade_MultiConsumeGap_V2_Ok {
+    uint256 a;
+    uint256 b;
+}
+
+contract StorageUpgrade_EndGap_V1a {
+    uint256 a;
+    uint256[1] private __gap;
+}
+
+contract StorageUpgrade_EndGap_V1b is StorageUpgrade_EndGap_V1a {
+    uint256[2] private __gap;
+}
+
+contract StorageUpgrade_EndGap_V1c is StorageUpgrade_EndGap_V1b {
+    uint256 b;
+}
+
+contract StorageUpgrade_EndGap_V2a {
+    uint256 a;
+    uint256[1] small;
+}
+
+contract StorageUpgrade_EndGap_V2b is StorageUpgrade_EndGap_V2a {
+    uint256 inserted1;
+    uint256 inserted2;
+    uint256 b;
+}
+
+contract StorageUpgrade_EndGap_Uint128_V1a {
+    uint128 a;
+    uint256[1] private __gap;
+}
+
+contract StorageUpgrade_EndGap_Uint128_V1b is StorageUpgrade_EndGap_Uint128_V1a {
+    uint256[2] private __gap;
+}
+
+contract StorageUpgrade_EndGap_Uint128_V1c is StorageUpgrade_EndGap_Uint128_V1b {
+    uint128 b;
+}
+
+contract StorageUpgrade_EndGap_Uint128_V2a {
+    uint128 a;
+    uint128[1] small;
+}
+
+contract StorageUpgrade_EndGap_Uint128_V2b is StorageUpgrade_EndGap_Uint128_V2a {
+    uint128[4] medium;
+    uint128 b;
+}
+
+contract StorageUpgrade_Uint256Gap_Address_V1 {
+    address a;
+    address b;
+    uint256[48] __gap;
+    address z;
+}
+
+contract StorageUpgrade_Uint256Gap_Address_V2 {
+    address a;
+    address b;
+    address c;
+    uint256[47] __gap;
+    address z;
+}
+
+contract StorageUpgrade_AddressGap_V1 {
+    address a;
+    address b;
+    address[48] __gap;
+    address z;
+}
+
+contract StorageUpgrade_AddressGap_V2 {
+    address a;
+    address b;
+    address c;
+    address[47] __gap;
+    address z;
+}
+
+contract StorageUpgrade_Uint128Gap_V1 {
+    uint128 a;
+    uint128[49] __gap;
```

Comment:
Its nice if we support that, but gaps made of array that are not `bytes32[]` or `uint256[]` are not essential to support

Reply:
Agree, although the implementation does not treat different types in any special way so they should all be supported.

Added tests for `bytes32[]` gaps.